File: Algorithms_python/77_Combinations.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        def dfs(i: int):
            if len(t) == k:
                ans.append(t[:])
                return
            if i > n:
                return
            for j in range(i, n + 1):
                t.append(j)
                dfs(j + 1)
                t.pop()
                
        ans = []
        t = []
        dfs(1)
        return ans
'''
DFS,每次扩展一位，往下深挖，然后移除最后一位，这样就能保证能出所有可能性
更新：简化了dfs的逻辑，现在只用传一个参数i
'''

File: Algorithms_C++/77_Combinations.cpp
class Solution {
public:
    vector<vector<int>> ans;
    vector<int> t;

    void dfs(int i, int n, int k){
        if (t.size() == k){
            ans.emplace_back(t);
            return;
        }
        if (i > n){
            return;
        }
        t.emplace_back(i);
        dfs(i+1, n, k);
        t.pop_back();
        dfs(i+1, n, k);
    }

    vector<vector<int>> combine(int n, int k) {
        dfs(1, n, k);
        return ans;
    }
};